### ISSUE TITLE
Crash on PMKWhen on failed promise bridged from Swift to Obj-C

### ISSUE THREAD

**Issue by PerishableDave** [mxcl/PromiseKit#392]
Steps to reproduce:

1. Create a promise in Swift that fails throwing an ErrorType enum.
2. Add the promise to a PMKWhen in Objective-C

Cause:

PMKWhen uses helper method `NSErrorSuplement` to pass the failed value from the failed promise into the joined promise. `NSErrorSuplement` takes the error and creates another error with the same class as the thrown error. Since the error was originally thrown from Swift, it is bridged by the internal class `_SwiftNativeNSError`, which cannot be instantiated directly.

**Comment by PerishableDave** [comment]
The comment in NSErrorSuplement mentions that wrapping the original error with NSUnderlyingError is a todo. If thats something the maintainers would still like to do I can do the work to make the changes.

**Comment by nathanhosselton** [comment]
#393 looks like what I would expect the TODO intended. However, it seems to me that this would be API breaking, so I cannot apply it to master as-is without a major version bump.

Though it would add complexity, I think ideally the interim fix would check if initialization of the `[err class]` returns `nil` and only in that case utilize this new solution. It's still not ideal, but at least this way we aren't breaking error handlers in our users' codebases and we don't have to wait for or force a major version bump to fix this issue.

Since this is such a special case, I think it would also be warranted to print to the console when execution is forced to follow the new path, letting the user know that they need to use the `NSUnderlyingError` key of our userInfo dictionary to access their own. This message, along with the old implementation, would then be removed with 4.0.0 where we would document the new process for accessing the underlying error of a failed PMKWhen.

An alternative would simply be to crash when we can't reproduce the error object and inform the user that PMKWhen is not supported in Swift. I don't like this as much, but it's the easiest workaround.

**Comment by PerishableDave** [comment]
The interim fix sounds fine. We'd have to check the class to see if it's _SwiftNativeNSError, since trying to instantiate it produces an exception rather than returning nil. I'll make the changes so long as you're ok with it.

**Comment by nathanhosselton** [comment]
Ah, well that's not really nice. I dislike the idea of checking for the specific class, even if it is likely to catch all of the cases we want. Perhaps we should catch the exception then, since this crash is our own fault, not the library consumer's.

**Comment by nathanhosselton** [comment]
The above would require a little extra syntax for it to work within the macro:

``` objective-c
#define NSErrorSupplement(_err, supplements) ({ \
    NSError *err = _err; \
    id userInfo = err.userInfo.mutableCopy ?: [NSMutableArray new]; \
    [userInfo addEntriesFromDictionary:supplements]; \
    id whenErr; \
    @try { \
        whenErr = [[[err class] alloc] initWithDomain:err.domain code:err.code userInfo:userInfo]; \
    } \
    @catch (id exception) { \
        whenErr = [[NSError alloc] initWithDomain:err.domain code:err.code userInfo:userInfo]; \
    } \
    whenErr;\
})
```

**Comment by nathanhosselton** [comment]
Neither option is especially nice and they're both temporary. So choose whichever you think is best.

**Comment by PerishableDave** [comment]
Hrm, actually none of those solutions work. Trying to instantiate _SwiftNativeNSError throws EXC_BAD_ACCESS and can't be caught. We also can't check for the class since _SwiftNativeNSError isn't public. I'm trying to think of other options, but don't have any other ideas at the moment. Do you want to leave it broken till 4.0 or disregard it entirely? Anyways, whatever you want to do.

For reference:
[_SwiftNativeNSError](https://github.com/apple/swift/blob/master/stdlib/public/runtime/ErrorObject.mm#L41)

**Comment by nathanhosselton** [comment]
We can check for a class by string if we know the class name (and we do):

```objective-c
if ([NSStringFromClass([error class]) isEqualToString:"_SwiftNativeNSError"]) {
    // instantiate new error and print instructions for accessing to the console
} else {
    // carry on
};
```
If we must hard code the class name then so be it. It's probably safe.

**Comment by mxcl** [comment]
Merged #393
